### PR TITLE
bpo-36817: Add f-string debugging using '='.

### PR DESCRIPTION
**Pull request by ericvsmith**
https://bugs.python.org/issue36817

### CODE REVIEW

File: Doc/whatsnew/3.8.rst [hunk 1]
```diff
@@ -148,6 +148,20 @@ extensions compiled in release mode and for C extensions compiled with the
 stable ABI.
 (Contributed by Victor Stinner in :issue:`36722`.)
 
+f-strings now support !d for quick and dirty  debugging
+-------------------------------------------------------
+
+Add ``!d`` conversion specifier to f-strings. ``f'{expr!d}'`` expands
```
Comment (tirkarthi) [inline Doc/whatsnew/3.8.rst]: Since this is an alternative to PR that implements `!d` I am just adding a note to update this in case this PR is merged to use `=` in 3.8 .

File: Parser/Python.asdl [hunk 2]
```diff
@@ -76,7 +76,7 @@ module Python
          -- x < 4 < 3 and (x < 4) < 3
          | Compare(expr left, cmpop* ops, expr* comparators)
          | Call(expr func, expr* args, keyword* keywords)
-         | FormattedValue(expr value, int? conversion, expr? format_spec)
+         | FormattedValue(expr value, int? conversion, expr? format_spec, string? expr_text)
```
Comment (serhiy-storchaka) [inline Parser/Python.asdl:79]: It is not needed and I prefer to not add redundant attributes. Instead you can add the Constant item just before FormattedValue into JoinedStr.

File: Lib/test/test_future.py [hunk 3]
```diff
@@ -250,11 +250,21 @@ def test_annotations(self):
         eq('str or None if sys.version_info[0] > (3,) else str or bytes or None')
         eq("f'f-string without formatted values is just a string'")
         eq("f'{{NOT a formatted value}}'")
-        eq("f'some f-string with {a} {few():.2f} {formatted.values!r}'")
-        eq('''f"{f'{nested} inner'} outer"''')
-        eq("f'space between opening braces: { {a for a in (1, 2, 3)}}'")
-        eq("f'{(lambda x: x)}'")
-        eq("f'{(None if a else lambda x: x)}'")
+        eq("f'some f-string with {a!f} {few()!f:.2f} {formatted.values!r}'")
```
Comment (serhiy-storchaka) [inline Lib/test/test_future.py]: Why these tests were changed? Can you preserve the existing tests and add new tests?

File: Lib/test/test_fstring.py [hunk 4]
```diff
@@ -1049,6 +1064,36 @@ def test_backslash_char(self):
         self.assertEqual(eval('f"\\\n"'), '')
         self.assertEqual(eval('f"\\\r"'), '')
 
+    def test_debug_conversion(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_fstring.py:1067]: Please add also tests for expressions containing "=": with operators "==", "!=", "<=", ">=", ":=" and string literals containing "=" and for format spec containing "=" (`f'{x:=10}'` where `x` is a number).

File: Parser/Python.asdl (hunk 2 above)
Comment (ericvsmith) [reply Parser/Python.asdl:79]: It's used in ast_unparse.c to know if the = needs to be added back in. It's not strictly necessary, of course. It depends how you feel about the fidelity required in string annotations.

File: Lib/test/test_future.py (hunk 3 above)
Comment (ericvsmith) [reply Lib/test/test_future.py]: I'll double check. These might (or might not) be left over from the !d version. The ast unparser code has been through a number of revisions.

File: Lib/test/test_fstring.py (hunk 4 above)
Comment (ericvsmith) [reply Lib/test/test_fstring.py:1067]: Good idea.

File: Parser/Python.asdl (hunk 2 above)
Comment (serhiy-storchaka) [reply Parser/Python.asdl:79]: I think it if fine if `f'{foo=}'` be transformed into `f'foo={foo!r}'`. Think about this as a kind of macros.

Comment (ericvsmith) [reply Parser/Python.asdl:79]: I discussed this very point with Łukasz yesterday. I think his point is that he'd like to not lose the information in the unparser. Admittedly it's not perfect, and it probably can't ever be without a ton of work and additions to the AST.

File: Lib/test/test_future.py (hunk 3 above)
Comment (ericvsmith) [reply Lib/test/test_future.py]: They were changed because at one point the AST didn't have enough information to know whether !f had been used in the original source, so the unparser always added !f (since it was the default). This is another case where the exact fidelity of the annotations unparser is an issue.

But in the current implementation, the !f isn't added unless it's explicitly provided, so I can undo these changes.

Comment (ericvsmith) [reply Lib/test/test_future.py]: I restored these. Thanks for catching it.

File: Lib/test/test_fstring.py (hunk 4 above)
Comment (ericvsmith) [reply Lib/test/test_fstring.py:1067]: I've added these tests.

**Comment by larryhastings** [comment]
The two failed tests both failed in the same way: Lib/test/test_fstring.py failed a nanny check for whitespace.  "make patchcheck" will rub your nose in it.

File: Lib/test/test_fstring.py [hunk 5]
```diff
@@ -1049,6 +1064,90 @@ def test_backslash_char(self):
...
+        self.assertEqual(f'{0>=1}', 'False')
+        self.assertEqual(f'{(x:="5")}', '5')
+        self.assertEqual(x, '5')
+        self.assertEqual(f'{(x:=5)}', '5')
+        self.assertEqual(x, 5)
+        self.assertEqual(f'{"="}', '=')
+
+        x = 20
+        # This isn't an assignment expression, it's 'x', with a format
+        # spec of '=10'.
```
Comment (larryhastings) [inline Lib/test/test_fstring.py]: It's fussy and optional, but as long as you're telling the reader it's not an assignment expression, you might add:

(Assignment expressions are supported in f-strings but they must be enclosed in parentheses.)

File: Lib/test/test_fstring.py [hunk 6]
```diff
@@ -1049,6 +1064,90 @@ def test_backslash_char(self):
...
+        tenπ = 31.4
+        self.assertEqual(f'{tenπ=:.2f}', 'tenπ=31.40')
+
+        # Also test with Unicode in non-identifiers.
+        self.assertEqual(f'{"Σ"=}', '"Σ"=\'Σ\'')
+
+        # Make sure nested fstrings still work.
+        self.assertEqual(f'{f"{3.1415=:.1f}":*^20}', '*****3.1415=3.1*****')
+
+        # Make sure text before and after !d works correctly.
```
Comment (larryhastings) [inline Lib/test/test_fstring.py]: Fix mention of "!d"

File: Misc/NEWS.d/next/Core and Builtins/2019-05-02-11-48-08.bpo-36774.ZqbJ1J.rst [hunk 7]
```diff
@@ -0,0 +1,7 @@
+Add a ``=`` conversion feature to f-strings. This can precede ``!s``,
```
Comment (larryhastings) [inline Misc/NEWS.d/next/Core and Builtins/2019-05-02-11-48-08.bpo-36774.ZqbJ1J.rst]: I think this text is a remnant of when it was "!d".  Back then it was a new "conversion"; now it's its own separate thing and is *not* a conversion.  Maybe you're better off choosing a different term here, just to avoid muddying the waters?  How about this for a first sentence:

    Add a ``=`` feature to f-strings for debugging.

File: Misc/NEWS.d/next/Core and Builtins/2019-05-02-11-48-08.bpo-36774.ZqbJ1J.rst [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+``!r``, or ``!a``. It produces the text of the expression, followed by
+an equal sign, followed by the repr of the value of the expression. So
+``f'{3*9+15=}'`` would be equal to the string ``'3*9+15=42'``.  If
+``=`` is specified, the default conversion is set to ``!r``.  Added a
+``!f`` conversion to switch back to the previous format behavior, if
```
Comment (larryhastings) [inline Misc/NEWS.d/next/Core and Builtins/2019-05-02-11-48-08.bpo-36774.ZqbJ1J.rst]: "!f" is now removed, and the default is now repr-unless-there-is-a-format-spec etc etc.

File: Python/ast.c [hunk 9]
```diff
@@ -5016,7 +5017,12 @@ fstring_find_expr(const char **str, const char *end, int raw, int recurse_lvl,
     const char *expr_end;
     expr_ty simple_expression;
     expr_ty format_spec = NULL; /* Optional format specifier. */
-    int conversion = -1; /* The conversion char. -1 if not specified. */
+    int conversion = -1; /* The conversion char.  Use default if not
+                            specified, or !r if using = and no format
+                            spec. */
+    int equal_conversion = 0; /* Are we using the = conversion? */
```
Comment (larryhastings) [inline Python/ast.c]: Not a "conversion" anymore.  Rename variable?  Change comment?  (Too fussy?)

File: Python/ast.c [hunk 10]
```diff
@@ -5170,7 +5192,20 @@ fstring_find_expr(const char **str, const char *end, int raw, int recurse_lvl,
        conversion or format_spec. */
     simple_expression = fstring_compile_expr(expr_start, expr_end, c, n);
     if (!simple_expression)
-        return -1;
+        goto error;
+
+    /* Check for = conversion. */
```
Comment (larryhastings) [inline Python/ast.c]: Just noting all the places you call '=' a *conversion* in case you want to use a different term.  Here's another!

File: Python/ast.c [hunk 11]
```diff
@@ -5170,7 +5192,20 @@ fstring_find_expr(const char **str, const char *end, int raw, int recurse_lvl,
...
+    if (**str == '=') {
+        *str += 1;
+        equal_conversion = 1;
+
+        /* Skip over ASCII whitespace.  No need to test for end of string
```
Comment (larryhastings) [inline Python/ast.c:5203]: IIUC the comment is misleading.  If we have the malformed f-string

    f'{x='

then at this point there *isn't* "a } somewhere ahead".  Py_ISSPACE('\0') will return 0, so you'll exit the loop there, and you'll complain that you didn't stop on a '}'.  So the code works fine.  But not because a } is guaranteed--it's not like you've already validated the f-string or anything.

File: Python/ast_unparse.c [hunk 12]
```diff
@@ -655,6 +655,11 @@ append_formattedvalue(_PyUnicodeWriter *writer, expr_ty e, bool is_format_spec)
     }
     Py_DECREF(temp_fv_str);
 
+    if (e->v.FormattedValue.expr_text) {
+        /* Use the = debug conversion. */
```
Comment (larryhastings) [inline Python/ast_unparse.c]: Another place you call it a "conversion".

File: Python/compile.c [hunk 13]
```diff
@@ -3946,8 +3946,8 @@ compiler_formatted_value(struct compiler *c, expr_ty e)
     /* Our oparg encodes 2 pieces of information: the conversion
        character, and whether or not a format_spec was provided.
 
-       Convert the conversion char to 2 bits:
-       None: 000  0x0  FVC_NONE
+       Convert the conversion char to 3 bits:
+       !f  : 000  0x0  FVC_NONE   The default if nothing specified.
```
Comment (larryhastings) [inline Python/compile.c]: "!f" has been removed, is this still relevant?

**Review by larryhastings — COMMENTED** [review]
Minor nonsense, mostly about comments.  I didn't mark this as "Approve" just because I wanted to see what you had to say in reply; if you tell me it's all fine then I'll probably just Approve it then.

**Comment by ericvsmith** [comment]
I fixed the whitespace problem. I kept fighting emacs over that, and couldn't figure out what problem was. Thanks, Larry and patchcheck.

I'm not sure which tests you're referencing that are failing. All of the tests pass for me on macOS.

File: Python/ast.c (hunk 9 above)
Comment (ericvsmith) [reply Python/ast.c]: I'm just going to leave this one, for lack of a better name. I'll fix the comment.

**Comment by larryhastings** [comment]
The test I was referring to were CI tests.  Azure and something else were both complaining about the whitespace, and with your latest update now they're not.

File: Lib/test/test_fstring.py [hunk 14]
```diff
@@ -1147,6 +1148,15 @@ def __repr__(self):
         self.assertEqual(f'{C()=:x}', 'C()=FORMAT-x')
         self.assertEqual(f'{C()=!r:*^20}', 'C()=********REPR********')
 
+    def test_walrus(self):
+        x = 20
+        # This isn't an assignment expression, it's 'x', with a format
+        # spec of '=10'.  See test_walrus: you need to use parens.
```
Comment (larryhastings) [inline Lib/test/test_fstring.py]: def test_walrus(self):
        ...
        # See test_walrus:

File: Python/ast.c [hunk 15]
```diff
@@ -5020,7 +5020,7 @@ fstring_find_expr(const char **str, const char *end, int raw, int recurse_lvl,
     int conversion = -1; /* The conversion char.  Use default if not
                             specified, or !r if using = and no format
                             spec. */
-    int equal_conversion = 0; /* Are we using the = conversion? */
+    int equal_conversion = 0; /* Are we using the = feature? */
```
Comment (larryhastings) [inline Python/ast.c]: So you still call the variable "equal_conversion"?  I figured you'd rename it.  "equal_debugging" maybe?

**Review by larryhastings — CHANGES_REQUESTED** [review]
OH WHY DO PEOPLE COMPLAIN ABOUT THIS PIDDLING LITTLE STUFF

MY LIFE

**Review by larryhastings — APPROVED** [review]
